dx/compare.go

```diff
@@ -0,0 +1,350 @@
+package dx
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"reflect"
+	"text/tabwriter"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+)
+
+// NewCompareCommand initializes a new compare command for dx.
+func NewCompareCommand() *cobra.Command {
+	compareCmd := &cobra.Command{
+		Use:   "compare",
+		Short: "compare two dx solo results",
+		Long:  `Compare two result files generated by dx solo command.`,
+		Args: func(cmd *cobra.Command, args []string) error {
+			return validateComparisonArgs(args)
+		},
+		Run: func(cmd *cobra.Command, args []string) {
+			if err := ExecuteComparison(args[0], args[1]); err != nil {
+				fmt.Printf("%+v", err)
+				os.Exit(1)
+			}
+		},
+	}
+
+	return compareCmd
+}
+
+// validateCompareArgs validates that the args passed to compare command has length exactly 2
+// and are valid filenames.
+func validateComparisonArgs(args []string) error {
+	if len(args) != 2 {
+		return errors.New("need exactly two files to compare")
+	}
+	fileExists, err := checkFileExists(args[0])
+	if err != nil {
+		return errors.Wrapf(err, "error verifying file %s exists", args[0])
+	}
+	if !fileExists {
+		return errors.Errorf("%s does not exist", args[0])
+	}
+	fileExists, err = checkFileExists(args[1])
+	if err != nil {
+		return errors.Wrapf(err, "error verifying file %s exists", args[1])
+	}
+	if !fileExists {
+		return errors.Errorf("%s does not exist", args[1])
+	}
+	return nil
+}
+
+// Comparison struct contains the information of a comparison. RunTime is the total time it took for the run to complete.
+// The TotalTime is the total of all the individual times of each operation, which may have been running in separate goroutines.
+type Comparison struct {
+	Type           string
+	RunTime1       time.Duration
+	RunTime2       time.Duration
+	RunTimeDelta   float64
+	TotalTime1     time.Duration
+	TotalTime2     time.Duration
+	TotalTimeDelta float64
+	ThreadCount1   int
+	ThreadCount2   int
+	Accuracy       float64
+	Size           int64
+}
+
+// ExecuteComparison executes a comparison on the two files.
+func ExecuteComparison(file1, file2 string) error {
```

Comment:
the validator actually checks if the file is regular and not a directory, but the error being returned prints out the stack trace by default (and so looks like a panic). I think the bad data is handled similarly in the `ExecuteComparison` command. I'll change it so that the stack trace is only printed out when `--verbose` is set to true.